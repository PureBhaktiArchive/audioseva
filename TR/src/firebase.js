import { initializeApp } from 'firebase/app';
import { getAuth, getRedirectResult } from 'firebase/auth';

export const firebaseApp = initializeApp({
  ...JSON.parse(import.meta.env.VITE_FIREBASE_CONFIG),
  // Patching the authDomain property according to https://firebase.google.com/docs/auth/web/redirect-best-practices
  // Option 1 in production, Option 3 in development. In both options, patching authDomain is needed.
  authDomain: window.location.host,
});

// Running this each time because who knows if there was a redirect.
getRedirectResult(getAuth());
