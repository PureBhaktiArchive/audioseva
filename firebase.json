{
  "database": {
    "rules": "./database.rules.json"
  },
  "storage": [
    {
      "target": "uploads",
      "rules": "./functions/storage.uploads.rules"
    }
  ],
  "hosting": [
    {
      "target": "main",
    "public": "frontend/dist",
    "ignore": ["firebase.json", "**/.*", "**/node_modules/**"],
    "rewrites": [
      {
        "source": "**",
        "destination": "/index.html"
      }
    ],
    "predeploy": [
      "npm --prefix \"$RESOURCE_DIR/..\" run build -- --mode $GCLOUD_PROJECT"
    ]
  },
    {
      "target": "CR",
      "public": "allotment",
      "ignore": ["firebase.json", "**/.*", "**/node_modules/**"],
      "headers": [
        {
          "source": "**/*.js",
          "headers": [
            {
              "key": "Cache-Control",
              "value": "max-age=0,no-cache"
            }
          ]
        },
        {
          "source": "404.html",
          "headers": [
            {
              "key": "Cache-Control",
              "value": "max-age=300"
            }
          ]
        }
      ]
    }
  ],
  "functions": {
    "predeploy": [
      "npm --prefix \"$RESOURCE_DIR\" run lint",
      "npm --prefix \"$RESOURCE_DIR\" run build"
    ]
  }
}
