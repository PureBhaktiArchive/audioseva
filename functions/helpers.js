const functions = require('firebase-functions');
let SibApiV3Sdk = require('sib-api-v3-sdk');
const sendInBlueSecretKey = functions.config().audioseva.send_in_blue.key;
const defaultClient = SibApiV3Sdk.ApiClient.instance;
let apiKey = defaultClient.authentications['api-key'];
apiKey.apiKey = sendInBlueSecretKey;



exports.checkValidMP3 = filePath => (filePath.startsWith("mp3/") && filePath.endsWith(".mp3"))


exports.removeFromDB = (db, dbPath) => {
    let ref = db.ref(dbPath);
    ref.remove()
        .then(() => console.log("Deleted."))
        .catch(error => console.log(error));
}


exports.sendEmail = (to, bcc, templateId, params) => {
    let apiInstance = new SibApiV3Sdk.SMTPApi();
    sendEmail = new SibApiV3Sdk.SendSmtpEmail();

    sendEmail = {
        to: [{ email: to }], // add a name prop if it fails to send
        bcc,
        templateId,
        params
    };

    apiInstance.sendTransacEmail(sendEmail).then(function(data) {
        console.log('API called successfully. Returned data: ');
        console.log(data);
        return 1;
    }, err => console.log(err)).catch(err => console.log(err));
}
