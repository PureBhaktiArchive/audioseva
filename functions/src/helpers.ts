const functions = require('firebase-functions');
let SibApiV3Sdk = require('sib-api-v3-sdk');
const sendInBlueSecretKey = functions.config().send_in_blue.key;
const defaultClient = SibApiV3Sdk.ApiClient.instance;
let apiKey = defaultClient.authentications['api-key'];
apiKey.apiKey = sendInBlueSecretKey;



export const checkValidMP3 = filePath => (filePath.startsWith("mp3/") && filePath.endsWith(".mp3"))

export const createMP3DBRef = (filePath, _module) => {
    const parts = filePath.split('/');
        
    const list = parts[1];
    const mp3 = parts[2];
    let file_name = mp3.slice(0, -4);

    return `/${_module}/files/${list}/${file_name}`;
}

export const extractListFromFilename = (fileName) => {
    return fileName.match(/^[^-]*[^ -]/g)[0];
}


export const storeFileNameToDB = (filePath, db, _module) => {
    if(checkValidMP3(filePath)) {
        let ref = db.ref(createMP3DBRef(filePath, _module));

        //checking if the file already exists in the RT db
        ref.child("status").once('value')
        .then(snapshot => {
            if(!snapshot.exists())
                ref.set({status: "Spare"});
            return 1;
        }).catch(err => console.log(err));
    }
}

export const removeFromDB = (db, dbPath) => {
    let ref = db.ref(dbPath);
    ref.remove()
        .then(() => console.log("Deleted."))
        .catch(error => console.log(error));
}


export const sendEmail = (to, bcc, templateId, params) => {
    let apiInstance = new SibApiV3Sdk.SMTPApi();
    let sendEmail = new SibApiV3Sdk.SendSmtpEmail();

    sendEmail = {
        to: [{ email: to }], // add a name prop if it fails to send
        bcc,
        templateId,
        params
    };

    apiInstance.sendTransacEmail(sendEmail).then(function(data) {
        // Message sending result
        console.log(data);
        return 1;
    }, err => console.log(err)).catch(err => console.log(err));
}
