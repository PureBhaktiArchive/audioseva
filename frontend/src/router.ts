/*
 * sri sri guru gauranga jayatah
 */

import firebase from "firebase/app";

import Vue from "vue";
import Router from "vue-router";

Vue.use(Router);

export const router = new Router({
  mode: "history",
  routes: [
    { path: "*", redirect: "/" },
    {
      path: "/login",
      meta: { guestOnly: true },
      component: () => import("@/views/Login.vue")
    },
    {
      path: "/",
      meta: { requireAuth: true },
      component: () => import("@/views/Dashboard.vue"),
      children: [
        {
          path: "",
          component: () => import("@/views/Home.vue")
        },
        {
          path: "cr/allot",
          component: () => import("@/views/CRAllotment.vue")
        },
        {
          path: "sqr/",
          component: () => import("@/views/SQR.vue"),
          meta: { activator: true, activatorName: "Sound Quality Reporting", menuIcon: "fas fa-headphones" },
          children: [
            { path: "",
              component: () => import("@/views/SQRFiles.vue"),
              meta: { menuItem: true}
            },
            {
              path: "allot",
              component: () => import("@/views/SQRAllotment.vue"),
              meta: { menuItem: true }
            },
            { path: "statistics",
              component: () => import("@/views/SQRFileStatistics.vue"),
              meta: { menuItem: true }
            }
          ]
        },
        {
          path: "se/",
          component: () => import("@/views/SE/SE.vue"),
          meta: { activator: true, activatorName: "Sound Engineering", menuIcon: "fas fa-music" },
          children: [
            { path: "", component: () => import("@/views/SE/Tasks.vue"), meta: { menuItem: true } },
            { path: "allot", component: () => import("@/views/SE/Allotment.vue"), meta: { menuItem: true } }
          ]
        },
        {
          path: "te/allot",
          component: () => import("@/views/TEAllotment.vue")
        },
        {
          path: "te/fc/allot",
          component: () => import("@/views/TFCAllotment.vue")
        },
        {
          path: "/qc/allot",
          component: () => import("@/views/QCAllotment.vue")
        }
      ]
    }
  ]
});

router.beforeEach((to, from, next) => {
  let currentUser = firebase.auth().currentUser;
  let requireAuth = to.matched.some(record => record.meta.requireAuth);
  let guestOnly = to.matched.some(record => record.meta.guestOnly);

  if (requireAuth && !currentUser)
    next({
      path: "/login",
      query: { redirect: to.fullPath }
    });
  else if (guestOnly && currentUser) next("/");
  else next();
});
