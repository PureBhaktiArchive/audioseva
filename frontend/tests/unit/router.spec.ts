import firebase from "firebase/app";
import { checkAuth, redirectSections } from "@/router";
import { mockClaims } from "./helpers";

describe("redirectSections", () => {
  let to: any;
  let from: any = {};
  let next: any;

  beforeEach(() => {
    next = jest.fn();
  });

  test.each`
    claims                           | toProps                                       | expectedPath
    ${{ TE: { editor: true } }}      | ${{ fullPath: "/te", toClaims: "TE.editor" }} | ${"/te/my"}
    ${{ TE: { coordinator: true } }} | ${{ fullPath: "/te" }}                        | ${"/te/tasks"}
  `(
    "should redirect to first available child route that matches claims $claims",
    async ({
      claims,
      expectedPath,
      toProps: { fullPath, toClaims = "TE.coordinator" }
    }) => {
      await mockClaims(claims);
      to = {
        fullPath: fullPath,
        meta: {
          activator: true,
          auth: { requireClaims: toClaims }
        }
      };
      await redirectSections(to, from, next);
      expect(next).toHaveBeenCalledWith(expectedPath);
    }
  );
});

describe("checkAuth", () => {
  let to: any;
  let from: any = {};
  let next: any;

  beforeEach(() => {
    next = jest.fn();
  });

  it("should redirect to login when requireClaims and no current user", async () => {
    (firebase.auth as any).mockImplementationOnce(() => ({
      currentUser: null
    }));
    to = {
      fullPath: "/te/tasks",
      matched: [{}, { meta: { auth: { requireClaims: "TE.coordinator" } } }]
    };
    await checkAuth(to, from, next);
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith({
      path: "/login",
      query: { redirect: to.fullPath }
    });
  });

  it("should redirect to / if guestOnly and currentUser", async () => {
    to = {
      matched: [{ meta: { auth: { guestOnly: true } } }]
    };
    await checkAuth(to, from, next);
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith("/");
  });

  it("should redirect to /login if requireAuth and no currentUser", async () => {
    (firebase.auth as any).mockImplementationOnce(() => ({
      currentUser: null
    }));
    to = {
      fullPath: "/sqr",
      matched: [{ meta: { auth: { requireAuth: true } } }]
    };
    await checkAuth(to, from, next);
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith({
      path: "/login",
      query: { redirect: to.fullPath }
    });
  });

  it("should allow route with no auth meta", async () => {
    to = {
      matched: [{ meta: {} }]
    };
    await checkAuth(to, from, next);
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith();
  });

  it("should not redirect on correct claims", async () => {
    (firebase.auth as any).mockImplementationOnce(() => ({
      currentUser: {
        getIdTokenResult: async () => {
          return {
            claims: { TE: { coordinator: true } }
          };
        }
      }
    }));

    to = {
      matched: [{ meta: { auth: { requireClaims: "TE.coordinator" } } }]
    };
    await checkAuth(to, from, next);
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith();
  });

  it("should redirect to / on bad custom claims", async () => {
    to = {
      matched: [{ meta: { auth: { requireClaims: "SQR.coordinator" } } }]
    };
    const from: any = {};
    const next: any = jest.fn();
    await checkAuth(to, from, next);
    expect(next).toHaveBeenCalledTimes(1);
    expect(next).toHaveBeenCalledWith("/");
  });
});
