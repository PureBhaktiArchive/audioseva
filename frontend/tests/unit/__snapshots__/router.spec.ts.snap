// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterRoutes filterRoutes 1`] = `
Array [
  Object {
    "meta": Object {
      "auth": Object {
        "requireClaims": Object {
          "SE": true,
        },
      },
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "meta": Object {
              "auth": Object {
                "requireClaims": Object {
                  "SE": true,
                },
              },
            },
          },
        ],
        "meta": Object {
          "activator": true,
          "auth": Object {
            "requireClaims": Object {
              "SQR": true,
            },
          },
        },
      },
    ],
    "meta": Object {
      "activator": true,
      "auth": Object {
        "requireClaims": Object {
          "CR": true,
        },
      },
    },
  },
]
`;

exports[`filterRoutesByClaims filterRoutesByClaims 1`] = `
Array [
  Object {
    "meta": Object {
      "auth": Object {
        "requireClaims": Object {
          "SE": true,
        },
      },
    },
  },
  Object {
    "children": Array [
      Object {
        "children": Array [
          Object {
            "meta": Object {
              "auth": Object {
                "requireClaims": Object {
                  "SE": true,
                },
              },
            },
          },
        ],
        "meta": Object {
          "activator": true,
          "auth": Object {
            "requireClaims": Object {
              "SQR": true,
            },
          },
        },
      },
    ],
    "meta": Object {
      "activator": true,
      "auth": Object {
        "requireClaims": Object {
          "CR": true,
        },
      },
    },
  },
]
`;
